design/vouching-contracts-v2.1.md

```diff
@@ -0,0 +1,178 @@
+
+`vouched` items will track the address of the object being vouched. In the case of EVM packages, it will be the deployment address of the ImplementationDirectory.
+
+Each `vouched` item should be composed of:
+- vouched address (immutable)
+- metadata URI+hash (immutable)
+- owner
+- total stake
+- stake per user
+
+## Registration and vouchings
+
+The vouching contract should expose the following interface for managing registrations and vouchings. Methods for deprecating or removing entries are out of scope.
+
+### Register
+
+```
+register(address vouched, uint256 amount, string metadataURI, bytes32 metadataHash)
+  requires vouch >= minStake
+  emits Register(uint256 indexed id, address indexed vouched, address owner, uint256 amount, string metadataURI, bytes32 metadataHash)
+```
+
+Generates a fresh ID and adds a new `vouched` item to the vouching contract, owned by the sender, with `amount` initial ZEP tokens sent by the sender. Requires vouching at least `minStake` tokens, which is a constant value.
+
+### Vouch
+
+```
+vouch(uint256 id, uint256 amount)
+  requires amount > 0
+  requires id exists
+  emits Vouch(uint256 indexed id, address indexed sender, uint256 amount)
+```
+
+Increases the vouch for the package identified by `id` by `amount` for `sender`.
+
+### Unvouch
+
+```
+unvouch(uint256 id, uint256 amount)
+  requires amount > 0
+  requires amount <= vouched(msg.sender, id)
+  requires vouched(msg.sender, id) - amount >= minStake if msg.sender == owner(id)
+  emits Unvouch(uint256 indexed id, address indexed sender, uint256 amount)
+```
+
+Decreases the vouch for the package identified by `id` by `amount` for `sender`. Note that if `sender` is the `vouched` owner, he cannot decrease his vouching under `minStake`.
+
+### Move
+
+```
+move(uint256 from, uint256 to, uint256 amount)
+  requires amount > 0
+  requires amount <= vouched(msg.sender, from)
+  requires packages from and to to exist
+  requires msg.sender != owner(from) || msg.sender == owner(to) || (vouched(msg.sender, id) - amount >= minStake)
+  emits Vouch(uint256 indexed id = from, address indexed sender, uint256 amount)
+  emits Unvouch(uint256 indexed id = to, address indexed sender, uint256 amount)
+```
+
+Moves vouching from a package to another by `amount`. Note that if the owner is moving stake to another package he owns as well, then the restriction of keeping at least minStake does not apply. This enables a maintainer to remove all vouching from a version to move it to a different one, without any penalties. It also enables any user to shift their vouching without any cost.
+
+### Transfer ownership
+
+```
+transfer(uint256 id, address newOwner, uint256 amount)
```

Comment:
`transfer` seems too loaded from `ERC20`, `transferOwnership` would clash with `Vouching is Ownable`, maybe `transferVouch`? Not too happy with this, but I'd look for a better name.